### DISCUSSION TITLE
Optimizations!

### DISCUSSION THREAD

**Discussion by McKrak** [Sandman13sq/DmrVBM-blender-to-gms2#15] [Ideas]
Hey, I've recently heard word of your animated model format, and after trying it out, I'm in love with it and how it works!

I've recently tried this and Snidr's SMF format, and overall, I like this a whole lot! However, there's one thing I wish could be improved. Optimization.

I'm not too sure if it's me, but updating animated models uses a lot of matrix_multiply() calls, which takes up a big percentage when profiled with the debugger. It runs much slower compared to Snidr's SMF model format. I believe the reasoning behind this is due to the fact that SMF does all most of the animation math in the shader, rather than in GML. I could be wrong, but I believe that is the reason why SMF runs faster.

At the same time though, I'd rather stick with using DmrVBM for multiple reasons. The Blender plugin is much better in a lot of ways. It gives me more control over what I want to export, the format, and animations to export. The SMF exporter doesn't work for me most of the time, and it also can't limit the exported armature to deform bones only.

It would also be nice if there was a way to export animations separately, as I plan to have a lot of animations for my game (sometimes shared between characters), and I'd rather have them as separate files rather than being loaded with the models.

**Comment by Sandman13sq** [comment]
I appreciate the feedback. I was ready to sunset the project and move on to other things, but recent attention may result in one more version. 

In making this project, I accepted that GM's VM performance isn't its strongest feature and therefore didn't expect anyone to use this with a high amount of complicated models.  I've written a slightly cleaner animation solution in other projects that builds transforms first then multiplies matrices once at the very end. If I can come up with other solutions via extensions or shaders I can implement options for that as well to push performance.

In the file dialog you can isolate exporting meshes and animations by toggling the `Meshes`, `Skeleton`, and `Animations` buttons at the top. You can set up an export for meshes and another for the animations. The buttons are there to give that flexibility.

**Comment by McKrak** [comment]
Yes, a shader-based approach would be really awesome to see! Perhaps it'll work in HTML as well? I'm not too sure if it's just me, but this doesn't seem to currently work in HTML5 exports.

I will say I am actually using this for a large-scale project of mine. I think the current set of features this extension provides is more than enough for what I need. It's just the fine-tuning that I feel is left here.
